In [1]:
import pandas as pd
import glob, os
import numpy as np
import matplotlib.pyplot as plt
import googlemaps
import gmaps
API_KEY = 'AIzaSyC-mLUTfYy6RwWdhnXQ5y4bYnPJLl_NRE0' 
gmaps_dev = googlemaps.Client(key=API_KEY) ## 
gmaps.configure(api_key=API_KEY) # 

In [2]:
ls crime_data/2017

apr17.xls*  dec17.xls*  jan17.xls*  jun17.xls*  may17.xls*  oct17.xls*
aug17.xls*  feb17.xls*  jul17.xls*  mar17.xls*  nov17.xls*  sep17.xls*


## combine all files into one dataframe

In [3]:
path = 'crime_data/2017'
all_files = glob.glob(os.path.join(path, "*.xls")) 

df_from_each_file = (pd.read_excel(f) for f in all_files)
df   = pd.concat(df_from_each_file, ignore_index=True)

WARNING *** file size (3078551) not 512 + multiple of sector size (512)
WARNING *** OLE2 inconsistency: SSCS size is 0 but SSAT size is non-zero


## We got some erros, lets see

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119727 entries, 0 to 119726
Data columns (total 13 columns):
# offenses      40378 non-null float64
Beat            119727 non-null object
Block Range     79349 non-null object
BlockRange      40378 non-null object
Date            119727 non-null object
Hour            119727 non-null int64
Offense Type    119727 non-null object
Offenses        79349 non-null float64
Premise         119240 non-null object
Street Name     79349 non-null object
StreetName      40378 non-null object
Suffix          119727 non-null object
Type            119727 non-null object
dtypes: float64(2), int64(1), object(10)
memory usage: 11.9+ MB


In [5]:
df.head()

# offenses   Beat Block Range BlockRange                 Date  Hour  \
0         1.0  10H10         NaN    200-299  2017-04-10 00:00:00    15   
1         1.0  10H10         NaN  2300-2399  2017-04-11 00:00:00    15   
2         1.0  10H10         NaN  2300-2399  2017-04-11 00:00:00    17   
3         1.0  10H10         NaN  4600-4699  2017-04-12 00:00:00     9   
4         1.0  10H10         NaN    100-199  2017-04-12 00:00:00    19   

  Offense Type  Offenses                                Premise Street Name  \
0     Burglary       NaN                     Residence or House         NaN   
1        Theft       NaN    Restaurant or Cafeteria Parking Lot         NaN   
2        Theft       NaN    Restaurant or Cafeteria Parking Lot         NaN   
3     Burglary       NaN  Miscellaneous Business (Non-Specific)         NaN   
4        Theft       NaN          Other, Unknown, or Not Listed         NaN   

  StreetName Suffix Type  
0    CLIFTON      -    -  
1      CANAL      -   ST  
2      CANAL      -   ST  
3      CANAL      -   ST  
4       ADAM      -   LN

## Lets create a copy

In [6]:
df1 = df.copy()
df1.head()

# offenses   Beat Block Range BlockRange                 Date  Hour  \
0         1.0  10H10         NaN    200-299  2017-04-10 00:00:00    15   
1         1.0  10H10         NaN  2300-2399  2017-04-11 00:00:00    15   
2         1.0  10H10         NaN  2300-2399  2017-04-11 00:00:00    17   
3         1.0  10H10         NaN  4600-4699  2017-04-12 00:00:00     9   
4         1.0  10H10         NaN    100-199  2017-04-12 00:00:00    19   

  Offense Type  Offenses                                Premise Street Name  \
0     Burglary       NaN                     Residence or House         NaN   
1        Theft       NaN    Restaurant or Cafeteria Parking Lot         NaN   
2        Theft       NaN    Restaurant or Cafeteria Parking Lot         NaN   
3     Burglary       NaN  Miscellaneous Business (Non-Specific)         NaN   
4        Theft       NaN          Other, Unknown, or Not Listed         NaN   

  StreetName Suffix Type  
0    CLIFTON      -    -  
1      CANAL      -   ST  
2      CANAL      -   ST  
3      CANAL      -   ST  
4       ADAM      -   LN

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119727 entries, 0 to 119726
Data columns (total 13 columns):
# offenses      40378 non-null float64
Beat            119727 non-null object
Block Range     79349 non-null object
BlockRange      40378 non-null object
Date            119727 non-null object
Hour            119727 non-null int64
Offense Type    119727 non-null object
Offenses        79349 non-null float64
Premise         119240 non-null object
Street Name     79349 non-null object
StreetName      40378 non-null object
Suffix          119727 non-null object
Type            119727 non-null object
dtypes: float64(2), int64(1), object(10)
memory usage: 11.9+ MB


## lets check our datatypes

In [8]:
df1.dtypes

# offenses      float64
Beat             object
Block Range      object
BlockRange       object
Date             object
Hour              int64
Offense Type     object
Offenses        float64
Premise          object
Street Name      object
StreetName       object
Suffix           object
Type             object
dtype: object

## Duplicate columns
We have duplicate columns with diffrent names
- "Block Range" & "BlockRange"
- "StreetName"  & "Street Name"
- "Offenses"  & "# offenses"

lets concat the columns remove nan values and reindex

In [9]:
df1['BlockRange'] = pd.concat([df1['Block Range'].dropna(), df1['BlockRange'].dropna()]).reindex_like(df)
df1['StreetName'] = pd.concat([df1['Street Name'].dropna(), df1['StreetName'].dropna()]).reindex_like(df)
df1['Offenses'] = pd.concat([df1['# offenses'].dropna(), df1['Offenses'].dropna()]).reindex_like(df)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119727 entries, 0 to 119726
Data columns (total 13 columns):
# offenses      40378 non-null float64
Beat            119727 non-null object
Block Range     79349 non-null object
BlockRange      40378 non-null object
Date            119727 non-null object
Hour            119727 non-null int64
Offense Type    119727 non-null object
Offenses        79349 non-null float64
Premise         119240 non-null object
Street Name     79349 non-null object
StreetName      40378 non-null object
Suffix          119727 non-null object
Type            119727 non-null object
dtypes: float64(2), int64(1), object(10)
memory usage: 11.9+ MB


## create a subdataframe with the columns that we want

In [11]:
df17 = df1[['Date','Beat','BlockRange','StreetName','Offense Type','Premise','Offenses','Hour']]

In [12]:
df17.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119727 entries, 0 to 119726
Data columns (total 8 columns):
Date            119727 non-null object
Beat            119727 non-null object
BlockRange      119727 non-null object
StreetName      119727 non-null object
Offense Type    119727 non-null object
Premise         119240 non-null object
Offenses        119727 non-null float64
Hour            119727 non-null int64
dtypes: float64(1), int64(1), object(6)
memory usage: 7.3+ MB


In [13]:
df17.tail()

Date Beat   BlockRange     StreetName        Offense Type  \
119722  09/29/2017  UNK  12600-12699    STAFFORD RD             Robbery   
119723  09/30/2017  UNK  10600-10699       BROOKLET               Theft   
119724  09/30/2017  UNK  12300-12399  WESTHEIMER RD             Robbery   
119725  09/30/2017  UNK  13200-13299    BEECHNUT ST  Aggravated Assault   
119726  11/10/2015  UNK    4900-4999       TEALGATE                Rape   

                                    Premise  Offenses  Hour  
119722            Road, Street, or Sidewalk       1.0     1  
119723            Road, Street, or Sidewalk       1.0    13  
119724  Restaurant or Cafeteria Parking Lot       1.0    19  
119725                   Highway or Freeway       1.0    17  
119726                   Residence or House       1.0    22

---
# Now we choose each column to see what is wrong with it

- Date
- Beat
- BlockRange
- StreetName
- Offense Type
- Premise
- Offenses
- Hour

## Date column
## convert Date column to datetime

In [14]:
df17['Date'] = pd.to_datetime(df17['Date'])

/home/salas/miniconda3/envs/sbc/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [15]:
df17.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119727 entries, 0 to 119726
Data columns (total 8 columns):
Date            119727 non-null datetime64[ns]
Beat            119727 non-null object
BlockRange      119727 non-null object
StreetName      119727 non-null object
Offense Type    119727 non-null object
Premise         119240 non-null object
Offenses        119727 non-null float64
Hour            119727 non-null int64
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 7.3+ MB


In [16]:
df17.head()

Date   Beat BlockRange StreetName Offense Type  \
0 2017-04-10  10H10    200-299    CLIFTON     Burglary   
1 2017-04-11  10H10  2300-2399      CANAL        Theft   
2 2017-04-11  10H10  2300-2399      CANAL        Theft   
3 2017-04-12  10H10  4600-4699      CANAL     Burglary   
4 2017-04-12  10H10    100-199       ADAM        Theft   

                                 Premise  Offenses  Hour  
0                     Residence or House       1.0    15  
1    Restaurant or Cafeteria Parking Lot       1.0    15  
2    Restaurant or Cafeteria Parking Lot       1.0    17  
3  Miscellaneous Business (Non-Specific)       1.0     9  
4          Other, Unknown, or Not Listed       1.0    19

## Reindex dataframe by date

- reindex the dataframe and sort the dates

In [17]:
df17 = df17.set_index('Date').sort_index(ascending=True)

In [18]:
df17.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 119727 entries, 1917-01-20 to 2017-12-31
Data columns (total 7 columns):
Beat            119727 non-null object
BlockRange      119727 non-null object
StreetName      119727 non-null object
Offense Type    119727 non-null object
Premise         119240 non-null object
Offenses        119727 non-null float64
Hour            119727 non-null int64
dtypes: float64(1), int64(1), object(5)
memory usage: 7.3+ MB


In [19]:
df17.head()

Beat BlockRange   StreetName Offense Type  \
Date                                                     
1917-01-20  20G20  2100-2199      GESSNER        Theft   
1917-02-20  17E20  9100-9199      FONDREN   Auto Theft   
1917-02-20  10H70  7500-7599      ARDMORE        Theft   
1917-04-16  18F20  1100-1199  UPTOWN PARK         Rape   
1917-12-18   9C20    500-599     PORTWALL        Theft   

                                          Premise  Offenses  Hour  
Date                                                               
1917-01-20                  Apartment Parking Lot       1.0    12  
1917-02-20              Road, Street, or Sidewalk       1.0    15  
1917-02-20                      Other Parking Lot       1.0    16  
1917-04-16                      Bar or Night Club       1.0    12  
1917-12-18  Miscellaneous Business (Non-Specific)       1.0    14

## Some odd dates
`DatetimeIndex: 119727 entries, 1917-01-20 to 2017-12-31`

lets create two dataframes:
- where the dates are from jan 1 17 - Dec 31 17
- any rows that dont fall from the previous value

In [20]:
# df17.loc['2017'] both give the same value rows
df2017 = df17.loc['2017-01-01':'2017-12-31']

In [21]:
df2017.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 118581 entries, 2017-01-01 to 2017-12-31
Data columns (total 7 columns):
Beat            118581 non-null object
BlockRange      118581 non-null object
StreetName      118581 non-null object
Offense Type    118581 non-null object
Premise         118107 non-null object
Offenses        118581 non-null float64
Hour            118581 non-null int64
dtypes: float64(1), int64(1), object(5)
memory usage: 7.2+ MB


In [22]:
# create a separate df with all the extra rows for later
trash_17 = df17.loc[:'2016']

In [23]:
trash_17.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1146 entries, 1917-01-20 to 2016-12-31
Data columns (total 7 columns):
Beat            1146 non-null object
BlockRange      1146 non-null object
StreetName      1146 non-null object
Offense Type    1146 non-null object
Premise         1133 non-null object
Offenses        1146 non-null float64
Hour            1146 non-null int64
dtypes: float64(1), int64(1), object(5)
memory usage: 71.6+ KB


In [24]:
df2017.head()

Beat   BlockRange    StreetName        Offense Type  \
Date                                                               
2017-01-01  20G80  14600-14699  BRIAR FOREST               Theft   
2017-01-01  19G50  11700-11799      WILCREST               Theft   
2017-01-01   3B50    7300-7399     IRVINGTON          Auto Theft   
2017-01-01   3B50    6500-6599         AMASA  Aggravated Assault   
2017-01-01   2A50    6700-6799         ARNOT               Theft   

                                             Premise  Offenses  Hour  
Date                                                                  
2017-01-01                     Apartment Parking Lot       1.0    12  
2017-01-01             Bar or Night Club Parking Lot       1.0    19  
2017-01-01                         Bar or Night Club       1.0    12  
2017-01-01                 Road, Street, or Sidewalk       1.0     2  
2017-01-01  Parks and Recreation, Zoo, Swimming Pool       1.0     9

## Beat column

- original unique values: 237
- clean vearsion : 120


we will use `.str.strip()` to remove the extra empty spaces

In [25]:
len(df2017.Beat.unique())

237

In [26]:
df2017.Beat.unique()

array(['20G80', '19G50', '3B50', '2A50', '6B40', '19G40', '3B10', '24C20',
       '21I60', '2A20', '2A10', '2A30', '3B30', '2A60', '24C30', '2A40',
       '21I50', '20G70', '1A50', '5F30', '20G10', '1A20', '5F20', '6B20',
       '1A30', '6B10', '5F40', '20G30', '4F20', '20G60', '4F10', '1A10',
       '6B30', '20G20', '20G50', '5F10', '4F30', '20G40', '7C30', '16E10',
       '17E30', '13D20', '10H10', '11H30', '10H40', '14D10', '8C30',
       '8C40', '17E20', '8C50', '17E10', '14D20', '12D20', '7C50',
       '18F50', '8C10', '18F20', '12D40', '12D70', '13D10', '18F10',
       '14D30', '17E40', '14D40', 'UNK', '14D50', '16E20', '9C40',
       '15E10', '15E20', '15E40', '15E30', '9C20', '8C60', '9C10',
       '16E40', '16E30', '7C20', '10H50', '19G20', '10H60', '19G10',
       '11H40', '11H20', '11H10', '18F60', '6B50', '12D10', '7C10',
       '10H20', '10H70', '10H30', '6B60', '19G30', '18F30', '10H80',
       '24C50', '12D50', '21I20', '21I10', '18F40', '7C40', '3B40',
       '8C20', '1

In [27]:
df2017['Beat'] = df2017['Beat'].str.strip()

/home/salas/miniconda3/envs/sbc/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [28]:
len(df2017.Beat.unique())

120

In [29]:
df2017.Beat.unique()

array(['20G80', '19G50', '3B50', '2A50', '6B40', '19G40', '3B10', '24C20',
       '21I60', '2A20', '2A10', '2A30', '3B30', '2A60', '24C30', '2A40',
       '21I50', '20G70', '1A50', '5F30', '20G10', '1A20', '5F20', '6B20',
       '1A30', '6B10', '5F40', '20G30', '4F20', '20G60', '4F10', '1A10',
       '6B30', '20G20', '20G50', '5F10', '4F30', '20G40', '7C30', '16E10',
       '17E30', '13D20', '10H10', '11H30', '10H40', '14D10', '8C30',
       '8C40', '17E20', '8C50', '17E10', '14D20', '12D20', '7C50',
       '18F50', '8C10', '18F20', '12D40', '12D70', '13D10', '18F10',
       '14D30', '17E40', '14D40', 'UNK', '14D50', '16E20', '9C40',
       '15E10', '15E20', '15E40', '15E30', '9C20', '8C60', '9C10',
       '16E40', '16E30', '7C20', '10H50', '19G20', '10H60', '19G10',
       '11H40', '11H20', '11H10', '18F60', '6B50', '12D10', '7C10',
       '10H20', '10H70', '10H30', '6B60', '19G30', '18F30', '10H80',
       '24C50', '12D50', '21I20', '21I10', '18F40', '7C40', '3B40',
       '8C20', '1

## BlockRange columnm

In [30]:
len(df2017.BlockRange.unique())  # dont touch yet

320

## StreetName column

- original 21584
- clean : 18672

In [31]:
len(df2017.StreetName.unique())

21507

In [32]:
df2017['StreetName'] = df2017['StreetName'].str.strip()

/home/salas/miniconda3/envs/sbc/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [33]:
len(df2017.StreetName.unique())

18602

## Offense Type column

In [35]:
df2017['Offense Type'].unique()

array(['Theft', 'Auto Theft', 'Aggravated Assault', 'Burglary', 'Robbery',
       'Rape', 'Murder', 'AutoTheft', 1, 'Rape                     ',
       'Aggravated Assault       ', 'AutoTheft                ',
       'Burglary                 ', 'Theft                    ',
       'Robbery                  ', 'Murder                   '],
      dtype=object)

In [36]:
len(df2017['Offense Type'].unique())

16

In [37]:
df2017['Offense Type'] = df2017['Offense Type'].str.strip()

/home/salas/miniconda3/envs/sbc/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [38]:
df2017['Offense Type'].unique()

array(['Theft', 'Auto Theft', 'Aggravated Assault', 'Burglary', 'Robbery',
       'Rape', 'Murder', 'AutoTheft', nan], dtype=object)

In [39]:
len(df2017['Offense Type'].unique())

9

### Auto Theft & AutoTheft 
they are both similar, lets replace the spaced value with the otherone

In [40]:
df2017 = df2017.replace('Auto Theft','AutoTheft')

In [42]:
df2017['Offense Type'].unique()

array(['Theft', 'AutoTheft', 'Aggravated Assault', 'Burglary', 'Robbery',
       'Rape', 'Murder', nan], dtype=object)

In [41]:
len(df2017['Offense Type'].unique())

8

## Premise Column

In [43]:
df2017.Premise.unique()

array(['Apartment Parking Lot', 'Bar or Night Club Parking Lot',
       'Bar or Night Club', 'Road, Street, or Sidewalk',
       'Parks and Recreation, Zoo, Swimming Pool', 'Residence or House',
       nan, 'Apartment', 'Driveway', 'Airport Terminal',
       'Other Parking Lot', 'Restaurant or Cafeteria',
       'Strip Business Center Parking Lot',
       'Grocery Store or Supermarket', 'Service or Gas Station',
       'Restaurant or Cafeteria Parking Lot',
       'Vehicle/Auto Sales/Lease/Auto Parts Store', 'Bank',
       'Train (Not Metro) Terminal', 'Department or Discount Store',
       'Commercial Building', 'Rental Storage Facility',
       'Hotel, Motel, Inn, Etc.', 'Hotel or Motel Parking Lot',
       'Mobile Home', 'Other, Unknown, or Not Listed', 'Bus Stop',
       'Grocery Store or Supermarket Parking Lot',
       'Miscellaneous Business (Non-Specific)',
       'Theatres, Dinner Theaters, Auditoriums', 'Nursing Home',
       'Convenience Store', 'Rehabilitation Center',
    

In [44]:
len(df2017.Premise.unique())

253

In [45]:
df2017['Premise'] = df2017['Premise'].str.strip()

In [46]:
len(df2017.Premise.unique())

133

In [47]:
df2017.Premise.unique()

array(['Apartment Parking Lot', 'Bar or Night Club Parking Lot',
       'Bar or Night Club', 'Road, Street, or Sidewalk',
       'Parks and Recreation, Zoo, Swimming Pool', 'Residence or House',
       nan, 'Apartment', 'Driveway', 'Airport Terminal',
       'Other Parking Lot', 'Restaurant or Cafeteria',
       'Strip Business Center Parking Lot',
       'Grocery Store or Supermarket', 'Service or Gas Station',
       'Restaurant or Cafeteria Parking Lot',
       'Vehicle/Auto Sales/Lease/Auto Parts Store', 'Bank',
       'Train (Not Metro) Terminal', 'Department or Discount Store',
       'Commercial Building', 'Rental Storage Facility',
       'Hotel, Motel, Inn, Etc.', 'Hotel or Motel Parking Lot',
       'Mobile Home', 'Other, Unknown, or Not Listed', 'Bus Stop',
       'Grocery Store or Supermarket Parking Lot',
       'Miscellaneous Business (Non-Specific)',
       'Theatres, Dinner Theaters, Auditoriums', 'Nursing Home',
       'Convenience Store', 'Rehabilitation Center',
    

## Offenses

In [48]:
df2017.Offenses.unique()  # dont touch yet

array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8., 10.])

## Hour column

In [49]:
len(df2017.Hour.unique())  #dont touch yet

25

## look at clean df

In [51]:
df2017.head()

Beat   BlockRange    StreetName        Offense Type  \
Date                                                               
2017-01-01  20G80  14600-14699  BRIAR FOREST               Theft   
2017-01-01  19G50  11700-11799      WILCREST               Theft   
2017-01-01   3B50    7300-7399     IRVINGTON           AutoTheft   
2017-01-01   3B50    6500-6599         AMASA  Aggravated Assault   
2017-01-01   2A50    6700-6799         ARNOT               Theft   

                                             Premise  Offenses  Hour  
Date                                                                  
2017-01-01                     Apartment Parking Lot       1.0    12  
2017-01-01             Bar or Night Club Parking Lot       1.0    19  
2017-01-01                         Bar or Night Club       1.0    12  
2017-01-01                 Road, Street, or Sidewalk       1.0     2  
2017-01-01  Parks and Recreation, Zoo, Swimming Pool       1.0     9

## save our df to csv for later

In [52]:
df2017.to_csv('crime_data/2017/clean_2017.csv', encoding='utf8')


In [53]:
ls crime_data/2017

apr17.xls*       dec17.xls*  jul17.xls*  may17.xls*  sep17.xls*
aug17.xls*       feb17.xls*  jun17.xls*  nov17.xls*
clean_2017.csv*  jan17.xls*  mar17.xls*  oct17.xls*
